nucypher/blockchain/eth/signers.py

```diff
@@ -251,102 +258,165 @@ def unlock_account(self, account: str, password: str, duration: int = None) -> b
+        """Read the keystore directory from the disk and populate accounts."""
+        try:
+            files = os.listdir(path=path)
+        except FileNotFoundError:
+            raise self.InvalidSignerURI(f'No such keystore directory "{path}"')
+        for keyfile in files:
+            account, key_metadata = self.__handle_keyfile(path=keyfile)
+            self.__keys[account] = key_metadata
+
+    @staticmethod
+    def __read_keyfile(path: str) -> tuple:
+        """Read an individual keystore key file from the disk"""
+        with open(path, 'r') as keyfile:
+            key_metadata = json.load(keyfile)
+        address = key_metadata['address']
+        return address, key_metadata
+
+    def __handle_keyfile(self, path: str) -> Tuple[str, dict]:
+        """
+        Read a single keystore file from the disk and return its decoded json contents then internally
+        cache it on the keystore instance. Raises InvalidKeyfile if the keyfile is missing or corrupted.
+        """
+        try:
+            address, key_metadata = self.__read_keyfile(path=path)
+        except FileNotFoundError:
+            error = f"No such keyfile '{path}'"
+            raise self.InvalidKeyfile(error)
+        except JSONDecodeError:
+            error = f"Invalid JSON in keyfile at {path}"
+            raise self.InvalidKeyfile(error)
+        except KeyError:
+            error = "Keyfile does not contain address field at '{path}'"
+            raise self.InvalidKeyfile(error)
+        else:
+            if not is_address(address):
+                raise self.InvalidKeyfile(f"'{path}' does not contain a valid ethereum address")
+            address = to_checksum_address(address)
+        return address, key_metadata
 
-        with open(path) as file_obj:
-            key = json.load(file_obj)
+    @validate_checksum_address
+    def __get_signer(self, account: str) -> LocalAccount:
+        """Lookup a known keystore account by its checksum address or raise an error"""
+        try:
+            return self.__signers[account]
+        except KeyError:
+            if account not in self.__keys:
+                raise self.UnknownAccount(account=account)
+            else:
+                raise self.AccountLocked(account=account)
 
-        address = key['address']
-        if not is_address(address):
-            raise ValueError(f"'{path}' does not contain a valid address")
+    #
+    # Public API
+    #
 
-        self.__key = key
-        self.__account = to_checksum_address(address)
-        self.__signer = None
+    @property
+    def path(self) -> str:
+        """Read only access to the keystore path"""
+        return self.__path
 
     @classmethod
     def from_signer_uri(cls, uri: str) -> 'Signer':
+        """Return a keystore signer from URI string i.e. keystore:///my/path/keystore """
         decoded_uri = urlparse(uri)
-
         if decoded_uri.scheme not in ('key', 'keystore') or decoded_uri.netloc:
```

Comment:
If the intention is to check if `decoded_uri.scheme` is in `decoded_uri.netloc` I don't think that's happening.

Reply:
Checking for bad input here - If the scheme is not `'key'` or `'ketstore'` OR there is some kind of URL-like non path provided, parsed into `netloc`.